lib/ofdm_synchronization_cvf_impl.cc

```diff
@@ -0,0 +1,286 @@
+              d_wait_for_NULL = true;
+            } else {
+              // we expect the start of a new symbol here
+              // correlation has to be calculated completely new, because of skipping samples before
+              delayed_correlation(&in[i], true);
+              //GR_LOG_DEBUG(d_logger, format("  New possible peak %d (i=%d)") % d_correlation_normalized_magnitude % i);
+              // check if there is really a peak
+              if (d_correlation_normalized_magnitude > 0.3) { //TODO: check if we are on right edge
+                d_frequency_offset_per_sample = std::arg(d_correlation) / d_fft_length; // in rad/s
+                //add_item_tag(0, nitems_written(0) + i, pmt::mp("on track"), pmt::from_float(d_frequency_offset_per_sample));
+                //GR_LOG_DEBUG(d_logger, format("in track symbol %d") % (d_symbol_count));
+              } else {
+                // no peak found -> out of track; search for next NULL symbol
+                d_wait_for_NULL = true;
+                GR_LOG_DEBUG(d_logger, format("Lost track at %d, switching ot acquisition mode (%d)") %d_symbol_count %
+                                       d_correlation_normalized_magnitude);
+                /*add_item_tag(0, nitems_written(0) + i, pmt::mp("Lost"),
+                             pmt::from_float(d_correlation_normalized_magnitude));*/
+              }
+            }
+          } else if (d_cyclic_prefix_length*0.75 <= d_symbol_element_count && d_symbol_element_count < d_cyclic_prefix_length*0.75 + d_symbol_length) {
+            // calculate the complex frequency correction value
+            float oi, oq;
+            // fixed point sine and cosine
+            int32_t angle = gr::fxpt::float_to_fixed(d_phase);
+            gr::fxpt::sincos(angle, &oq, &oi);
+            gr_complex fine_frequency_correction = gr_complex(oi, oq);
+            // set tag at next item if it is the first element of the first symbol
+            if (d_symbol_count == 0 && d_symbol_element_count == d_cyclic_prefix_length) {
+              //GR_LOG_DEBUG(d_logger, format("Set Tag %d")%(nitems_read(0)+i));
+              add_item_tag(0, nitems_written(0) + d_nwritten, pmt::mp("Start"),
+                           pmt::from_float(std::arg(d_correlation)));
+
+            }
+            // now we start copying one symbol length to the output
+            out[d_nwritten++] = in[i] * fine_frequency_correction;
+          }
+          d_symbol_element_count++;
+        }
+        /* fine frequency correction:
+         * The frequency offset estimation is done above with the correlation.
+         * The modulated frequency runs parralel to the whole input stream, including the cyclic prefix, and is mixed
+         * to the output symbol samples.
+         */
+        d_phase += d_frequency_offset_per_sample;
+        //place phase in [-pi, +pi[
+#define F_PI ((float)(M_PI))
+        d_phase = std::fmod(d_phase + F_PI, 2.0f * F_PI) - F_PI;
```

Comment:
it's totally possible that I made a modulo VOLK kernel. Not sure it's upstream, though.